src/main/host/syscall/socket.c

```diff
@@ -1145,6 +1156,38 @@ SysCallReturn syscallhandler_setsockopt(SysCallHandler* sys,
                 sys, socket_desc, optname, optvalPtr, optlen);
             break;
         }
+        case SOL_TCP: {
+            if (descriptor_getType((LegacyDescriptor*)socket_desc) != DT_TCPSOCKET) {
+                errcode = -ENOPROTOOPT;
+                break;
+            }
+
+            /* Shadow doesn't support nagle's algorithm, so shadow always behaves as if TCP_NODELAY
+             * is enabled. Some programs will fail if setsockopt(fd, SOL_TCP, TCP_NODELAY, &1,
+             * sizeof(int)) returns an error, so we treat this as a no-op for compatibility.
+             */
+            if (optname == TCP_NODELAY) {
+                if (optlen < sizeof(int)) {
+                    errcode = -EINVAL;
+                    break;
+                }
+
+                int enable = 0;
+                errcode = process_readPtr(sys->process, &enable, optvalPtr, sizeof(int));
+                if (errcode == 0) {
+                    if (enable) {
+                        // wants to enable TCP_NODELAY
+                        debug("Ignoring TCP_NODELAY");
+                    } else {
+                        // wants to disable TCP_NODELAY
+                        warning("Cannot disable TCP_NODELAY");
+                        errcode = -ENOPROTOOPT;
```

Comment:
Do we want this to be an error? Or should we just print the warning but then pretend like it worked so we don't cause other programs to fail?

Reply:
> Do we want this to be an error?

Do you mean to have shadow exit with an error message? Or just bump this from a warning message to error message? The other syscalls that return a linux error only print warning-level log messages.

> Or should we just print the warning but then pretend like it worked so we don't cause other programs to fail?

I haven't seen any programs try to disable `TCP_NODELAY` (probably because it's disabled by default on Linux). For enabling `TCP_NODELAY` I think a no-op is the correct behaviour in shadow, but for disabling `TCP_NODELAY` I think a no-op would be incorrect behaviour, and we should wait to see if any applications actually fail when `setsockopt` returns an error and then decide if we should add the incorrect behaviour.

Reply:
OK, you convinced me. I'm fine with keeping as-is. I think it does make more sense to return an error when they are _explicitly_ trying to disable in a way we don't support. We can always keep a look out if returning an error here causes any apps to fail.